docs/source/advanced/tpu.rst

```diff
@@ -191,7 +191,59 @@ set the 16-bit flag.
 
 Under the hood the xla library will use the `bfloat16 type <https://en.wikipedia.org/wiki/Bfloat16_floating-point_format>`_.
 
-----------------
+
+-----------------
+
+Weight Sharing/Tying
+--------------------
+Weight Tying/Sharing is a technique where in the module weights are shared among two or more layers.
+This is a common method to reduce memory consumption and is utilized in many State of the Art
+architectures today.
+
+PyTorch XLA requires these weights to be tied/shared after moving the model
+to the TPU device. To support this requirement Lightning provides a model hook which is
+called after the model is moved to the device. Any weights that require to be tied should
+be done in the `on_post_move_to_device` model hook. This will ensure that the weights
+among the modules are shared and not copied.
+
+PyTorch Lightning has an inbuilt check which verifies that the model parameter lengths
+match once the model is moved to the device. If the lengths do not match Lightning
+throws a warning message.
+
+Example:
+
+.. code-block:: python
+
+    from pytorch_lightning.core.lightning import LightningModule
+    from torch import nn
+    from pytorch_lightning.trainer.trainer import Trainer
+
+
+    class WeightSharingModule(LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.layer_1 = nn.Linear(32, 10, bias=False)
+            self.layer_2 = nn.Linear(10, 32, bias=False)
+            self.layer_3 = nn.Linear(32, 10, bias=False)
+            self.layer_3.weight = self.layer_1.weight # Weights will be copied on TPU
```

Comment:
I find this line confusing. Is it needed ?

Reply:
yes, after moving to TPU shared weights are copied independently on the XLA device and are no longer shared. We need to reattach those weights.
Ref: https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#xla-tensor-quirks

The reason for doing it on CPU too might be if using a GPU we don't need to do this but on TPU it is required. So better do it always so that all you need is to change `Trainer(gpus/tpu_cores=...)`

Reply:
The model in __init__ will work for gpu's and cpu's correctly. For tpu's we would have to include the weight sharing part within the `on_post_move_to_device` hook too.

Reply:
Yes, make sense. Let's replace `# Weights will be copied on TPU` by `TPU shared weights are copied independently on the XLA device and are no longer shared. However, it works fine for CPU and GPU.`